src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -458,54 +442,88 @@ int32_t NaClSysNameService(struct NaClAppThread *natp,
   return retval;
 }
 
-int32_t NaClSysDup(struct NaClAppThread *natp,
-                   int                  oldfd) {
+int32_t NaClSysDup(struct NaClAppThread *natp, int oldfd) {
   struct NaClApp  *nap = natp->nap;
-  int             retval;
+  int             ret, newfd;
   struct NaClDesc *old_nd;
 
-  NaClLog(3, "NaClSysDup(0x%08"NACL_PRIxPTR", %d)\n",
-          (uintptr_t) natp, oldfd);
-  old_nd = NaClGetDesc(nap, oldfd);
-  if (NULL == old_nd) {
-    retval = -NACL_ABI_EBADF;
+  NaClLog(1, "NaClSysDup(0x%08"NACL_PRIxPTR", %d)\n", (uintptr_t)natp, oldfd);
+
+  if (nap->fd > FILE_DESC_MAX) {
+    ret = -NACL_ABI_EBADF;
+    goto done;
```

Comment:
The gotos are to keep the error path code style consistent with the rest of the syscalls that use gotos because of their more complicated error handling.